### PR TITLE
Connect Kube gateway part 4: frontend

### PR DESCRIPTION
**Pull request by greedy52**
Related PR:
- Part 1: https://github.com/gravitational/teleport/pull/28312
- Part 2: https://github.com/gravitational/teleport/pull/28920
- Part 3: https://github.com/gravitational/teleport/pull/29214

### CODE REVIEW

**Comment by greedy52** [comment]
Note that this change does not include migrating from old tsh kube terminal to new terminal.

And since my first time doing these change, not sure how much testing/story I should add.

@ravicious @gzdunek one usage thing, when the kube terminal is open, there is no indication of what user should do. `KUBECONFIG=<xxx>` is automatically set but there is no instructions. Maybe somehow we can print a message in the terminal?

File: web/packages/teleterm/src/ui/DocumentGatewayKube/DocumentGatewayKube.tsx [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+    case 'error': {
+      return (
+        <Reconnect
+          kubeId={params.kubeId}
+          statusText={connectAttempt.statusText}
+          reconnect={createGateway}
+        />
+      );
+    }
```
Comment (greedy52) [inline web/packages/teleterm/src/ui/DocumentGatewayKube/DocumentGatewayKube.tsx:76]: not sure how these look. happy to improve them if anyone has ideas.

File: web/packages/teleterm/src/ui/ConnectMyComputer/DocumentConnectMyComputerSetup/DocumentConnectMyComputerSetup.tsx [hunk 2]
```diff
@@ -35,9 +35,8 @@ interface DocumentConnectMyComputerSetupProps {
 export function DocumentConnectMyComputerSetup(
   props: DocumentConnectMyComputerSetupProps
 ) {
-  const [step, setStep] = useState<'information' | 'agent-setup'>(
-    'information'
-  );
+  const [step, setStep] =
+    useState<'information' | 'agent-setup'>('information');
```
Comment (greedy52) [inline web/packages/teleterm/src/ui/ConnectMyComputer/DocumentConnectMyComputerSetup/DocumentConnectMyComputerSetup.tsx:39]: hmm.. i think these get updated when I run prettier

Comment (ravicious) [reply web/packages/teleterm/src/ui/ConnectMyComputer/DocumentConnectMyComputerSetup/DocumentConnectMyComputerSetup.tsx:39]: This was fixed by #29352.

**Comment by ravicious** [comment]
I won't have the time to review this today, but I still wanted to offer some help.

> one usage thing, when the kube terminal is open, there is no indication of what user should do. `KUBECONFIG=<xxx>` is automatically set but there is no instructions. Maybe somehow we can print a message in the terminal?

We should be able to do it. We used to execute arbitrary commands by printing to the PTY which was not the best idea. But, instead of writing to PTY, we can make Xterm.js think that the PTY emitted some data, even though it's just us and not the actual PTY.

This works:

```diff
diff --git a/web/packages/teleterm/src/sharedProcess/ptyHost/ptyProcess.ts b/web/packages/teleterm/src/sharedProcess/ptyHost/ptyProcess.ts
index 9814d52ff5..5cf443234f 100644
--- a/web/packages/teleterm/src/sharedProcess/ptyHost/ptyProcess.ts
+++ b/web/packages/teleterm/src/sharedProcess/ptyHost/ptyProcess.ts
@@ -76,6 +76,11 @@ export class PtyProcess extends EventEmitter implements IPtyProcess {
     this._setStatus('open');
     this.emit(TermEventEnum.Open);
 
+    this.emit(
+      TermEventEnum.Data,
+      'hey hello \r\nsome fake data is being emitted here \r\n'
+    );
+
     this._process.onData(data => this._handleData(data));
     this._process.onExit(ev => this._handleExit(ev));
   }
```

And since we emit before we add the `onData` listener, we're guaranteed that it won't conflict with data actually emitted by the PTY.

Look how `initCommand` was used (#26837). We could instead have an option such as `initMessage` and then pass it only for those new kube terminals.

The message could be even the same as `tsh kube login` outputs currently.

File: lib/teleterm/gateway/kube.go [hunk 3]
```diff
@@ -209,3 +210,15 @@ func (k *kube) writeKubeconfig(key *keys.PrivateKey, cas map[string]tls.Certific
 	})
 	return nil
 }
+
+func (k *kube) CLICommand() (*api.GatewayCLICommand, error) {
+	// TODO(greedy52) currently kube must implement CLICommand in order to pass
+	// Kube to CLICommandProvider. We should revisit gateway design/flows like
```
Comment (ravicious) [inline lib/teleterm/gateway/kube.go:216]: Could you add a test which captures this behavior? If I remove this method, the code compiles and both unit and integration tests pass.

File: web/packages/teleterm/src/services/pty/types.ts [hunk 4]
```diff
@@ -41,6 +41,10 @@ export type PtyServiceClient = {
 export type ShellCommand = PtyCommandBase & {
   kind: 'pty.shell';
   cwd?: string;
+  // env is a record of additional env variables that need to be set for the shell terminal and it
+  // will be merged with process env.
+  env?: Record<string, string>;
+  helpMsg?: string;
```
Comment (ravicious) [inline web/packages/teleterm/src/services/pty/types.ts:47]: Let's add a comment explaining what `helpMsg` is, with the focus on the fact that it's not either written nor read from the underlying PTY.

File: web/packages/teleterm/src/sharedProcess/api/protogen/ptyHostService_grpc_pb.d.ts [hunk 5]
```diff
@@ -1,19 +1,3 @@
-/**
- * Copyright 2023 Gravitational, Inc
```
Comment (ravicious) [inline web/packages/teleterm/src/sharedProcess/api/protogen/ptyHostService_grpc_pb.d.ts:2]: Those "hand-generated" files in this dir are weird, you need to run `make fix-license` now to add the license back.

I was supposed to move these files to the proto dir but I've never had the time to do this.

File: web/packages/teleterm/src/sharedProcess/api/proto/ptyHostService.proto [hunk 6]
```diff
@@ -36,6 +36,7 @@ message PtyCreate {
   reserved 6;
   reserved "init_command";
   google.protobuf.Struct env = 7;
+  string helpmsg = 8;
```
Comment (ravicious) [inline web/packages/teleterm/src/sharedProcess/api/proto/ptyHostService.proto:39]: Shouldn't it be help_msg?
```suggestion
  string help_msg = 8;
```

File: web/packages/teleterm/src/sharedProcess/ptyHost/ptyProcess.ts [hunk 7]
```diff
@@ -75,6 +75,9 @@ export class PtyProcess extends EventEmitter implements IPtyProcess {
 
     this._setStatus('open');
     this.emit(TermEventEnum.Open);
+    if (this.options.helpMsg) {
+      this.emit(TermEventEnum.Data, this.options.helpMsg);
```
Comment (ravicious) [inline web/packages/teleterm/src/sharedProcess/ptyHost/ptyProcess.ts:79]: Let's explain how we want this to be displayed in terminal emulator in the app and circumvent the underlying PTY at the same time.

File: web/packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts [hunk 8]
```diff
@@ -233,14 +233,16 @@ async function setUpPtyProcess(
...
+  if (doc.kind === 'doc.terminal_tsh_kube' || doc.kind === 'doc.gateway_kube') {
     ctx.usageService.captureProtocolUse(clusterUri, 'kube', doc.origin);
   }
 
   const openContextMenu = () => ctx.mainProcessClient.openTerminalContextMenu();
 
   const refreshTitle = async () => {
-    if (cmd.kind !== 'pty.shell') {
+    // TODO(ravicious): Enable updating cwd in doc.gateway_kube titles by
+    // moving title-updating logic to DocumentsService.
```
Comment (ravicious) [inline web/packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts:244]: I know I suggested that in one of the previous PRs, but let's expand the comment since I probably won't remember what I meant in six months. :D

> The logic behind updating the title should be encapsulated in a single place, so that useDocumentTerminal doesn't need to know the details behind the title of each document kind.

File: web/packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts [hunk 9]
```diff
@@ -389,6 +391,40 @@ function createCmd(
...
+      );
+    }
+    const helpMsg =
+      "Started local proxy for Kubernetes cluster '" +
+      gateway.targetName +
+      "' and environment variable 'KUBECONFIG' is updated to the kubeconfig provided by the local proxy.\r\n\r\n" +
+      "Try 'kubectl version' to test the connection.\r\n\r\n";
+
+    return {
+      ...doc,
```
Comment (ravicious) [inline web/packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts:419]: Spreading on the `...doc` that we do for regular `pty.shell` below is a bad practice because it's not clear what fields from `doc` are needed in `pty.shell`.

Fortunately, here we can just remove this line as AFAIK we directly pass all required fields anyway.

File: web/packages/teleterm/src/ui/services/workspacesService/documentsService/types.ts [hunk 10]
```diff
@@ -133,6 +133,14 @@ export interface DocumentGatewayCliClient extends DocumentBase {
   status: '' | 'connecting' | 'connected' | 'error';
 }
 
+export interface DocumentGatewayKube extends DocumentBase {
+  kind: 'doc.gateway_kube';
+  rootClusterId: string;
+  leafClusterId: string | undefined;
```
Comment (ravicious) [inline web/packages/teleterm/src/ui/services/workspacesService/documentsService/types.ts:139]: You don't need these two fields for this document type AFAIK.

File: web/packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts [hunk 11]
```diff
@@ -389,6 +391,40 @@ function createCmd(
...
+      kind: 'pty.shell',
```
Comment (ravicious) [inline web/packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts:420]: Do you remember if we talked about persisting cwd and what was the outcome of the discussion?

**Review by ravicious — COMMENTED** [review]
I reviewed as much as I could today, I'll continue the review tomorrow.

File: web/packages/teleterm/src/ui/DocumentGatewayKube/DocumentGatewayKube.tsx
Comment (ravicious) [inline web/packages/teleterm/src/ui/DocumentGatewayKube/DocumentGatewayKube.tsx:1]: The docs need to be updated to mention that running Connect is now required for the KUBECONFIG to work. I suppose it'd be best to do it in a separate PR.

https://github.com/gravitational/teleport/blob/b3d123ef18b54a523d3ccaee1aeb0d12c8458cb8/docs/pages/connect-your-client/teleport-connect.mdx#L92-L104

File: web/packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts [hunk 12]
```diff
@@ -389,6 +391,40 @@ function createCmd(
...
+      // command for a kube gateway.
+      throw new Error(
+        `No KUBECONFIG provided for gateway ${gateway.targetUri}`
+      );
+    }
+    const helpMsg =
+      "Started local proxy for Kubernetes cluster '" +
+      gateway.targetName +
+      "' and environment variable 'KUBECONFIG' is updated to the kubeconfig provided by the local proxy.\r\n\r\n" +
+      "Try 'kubectl version' to test the connection.\r\n\r\n";
```
Comment (ravicious) [inline web/packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts:416]:
```suggestion
    const helpMsg =
      `Started a local proxy for Kubernetes cluster "${gateway.targetName}".\r\n\r\n` +
      'The KUBECONFIG env var can be used with third-party tools as long as the proxy is running.\r\n' +
      'Close the proxy from Connections in the top left corner or by closing Teleport Connect.\r\n\r\n' +
      "Try 'kubectl version' to test the connection.\r\n\r\n";
```

I feel like for a regular user, this will do a better job at explaining how it all works.

**Review by ravicious — COMMENTED** [review]
I tested it with the old kube tabs and everything works as expected. Overall I really like how this works and I'm glad we didn't run into any unexpected roadblocks, which means we planned everything pretty well.

I also tested this on Windows, no problems there.

The only thing we should add is more deprecation comments, maybe linking to an issue where we describe what the next steps are to move away from the previous implementation of kube tabs, based on the discussion from https://github.com/gravitational/teleport/pull/28312#discussion_r1253214517. Could you take care of that?

<details>
<summary>Helper function to open old kube tab</summary>

Put this in `connectToKube.ts`, then restart the app, open console in devtools in Electron and execute the function with the URI of the kube, which you can inspect also in devtools when opening the table with kubes.

```typescript
window['connectToOldKube'] = async uri => {
  const ctx: IAppContext = window['teleterm'];
  const rootClusterUri = routing.ensureRootClusterUri(uri);
  const documentsService =
    ctx.workspacesService.getWorkspaceDocumentService(rootClusterUri);
  const kubeDoc = documentsService.createTshKubeDocument({
    kubeUri: uri,
    origin: 'resource_table',
  });
  const connection = ctx.connectionTracker.findConnectionByDocument(kubeDoc);

  await ctx.workspacesService.setActiveWorkspace(rootClusterUri);

  documentsService.add({
    ...kubeDoc,
    kubeConfigRelativePath:
      (connection && connection['kubeConfigRelativePath']) ||
      kubeDoc.kubeConfigRelativePath,
  });
  documentsService.open(kubeDoc.uri);
};
```
</details>

File: web/packages/teleterm/src/ui/DocumentGatewayKube/DocumentGatewayKube.tsx (hunk 1 above)
Comment (ravicious) [reply web/packages/teleterm/src/ui/DocumentGatewayKube/DocumentGatewayKube.tsx:75]: Looks good. 👌

**Comment by ravicious** [comment]
> And since my first time doing these change, not sure how much testing/story I should add.

As far as frontend tests are involved, I think you added an adequate amount of them. The terminal tabs don't really have proper integration tests anyway and I wouldn't expect you to set all of this up.

File: web/packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts (hunk 11 above)
Comment (greedy52) [reply web/packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts:420]: No, I don't recall we have talked about cwd. is it updated in `refreshTitle`?

File: web/packages/teleterm/src/ui/services/workspacesService/documentsService/types.ts (hunk 10 above)
Comment (greedy52) [reply web/packages/teleterm/src/ui/services/workspacesService/documentsService/types.ts:139]: I got an error here when removing there 😃 
https://github.com/gravitational/teleport/blob/e444901d32e76ccbb1e3c37d8aaa85eb2fca9c22/web/packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts#L219-L238

File: web/packages/teleterm/src/services/tshd/types.ts [hunk 13]
```diff
@@ -46,7 +46,7 @@ export interface Server extends apiServer.Server.AsObject {
 
 export interface Gateway extends apiGateway.Gateway.AsObject {
   uri: uri.GatewayUri;
-  targetUri: uri.DatabaseUri;
+  targetUri: uri.DatabaseUri | uri.KubeUri;
```
[the remark this answers is not in the document]
Comment (greedy52) [reply web/packages/teleterm/src/services/tshd/types.ts:49]: I rewrote the logic a bit since short is same as long, and target name requires different parsing for kube. not sure if there is an easier way.
```
      let targetName: string;
      let targetUser: string;
      let targetDesc: string;

      // Try to get target name and user from gateway object.
      if (gateway) {
        targetName = gateway.targetName;
        targetUser = gateway.targetUser;
      }

      // Try to get target name from target URI.
      if (!targetName) {
        targetName = routing.parseDbUri(targetUri)?.params['dbId'];
      }
      if (!targetName) {
        targetName = routing.parseKubeUri(targetUri)?.params['kubeId'];
      }

      if (targetUser) {
        targetDesc = `${targetName} as ${targetUser}`;
      } else {
        targetDesc = targetName;
      }
```

**Comment by greedy52** [comment]
I have added some comments for removing the old implementation in [30bb201](https://github.com/gravitational/teleport/pull/29376/commits/30bb201d26f4b8e30240e9e57c86bea68a1a880f). I followed the way it's done in go but is there a standard how to mark deprecate in js?

File: web/packages/teleterm/src/ui/DocumentGatewayKube/DocumentGatewayKube.tsx
Comment (greedy52) [reply web/packages/teleterm/src/ui/DocumentGatewayKube/DocumentGatewayKube.tsx:1]: sounds good. i will do a separate doc change on this. And likely need to hold off merging the doc change until this is released.

File: web/packages/teleterm/src/ui/services/workspacesService/documentsService/types.ts [hunk 14]
```diff
@@ -201,6 +211,13 @@ export type CreateGatewayDocumentOpts = {
   origin: DocumentOrigin;
 };
 
+export type CreateGatewayKubeDocumentOpts = {
```
Comment (gzdunek) [inline web/packages/teleterm/src/ui/services/workspacesService/documentsService/types.ts:214]: Unused.

File: web/packages/teleterm/src/ui/services/workspacesService/documentsService/documentsService.ts [hunk 15]
```diff
@@ -188,6 +190,21 @@ export class DocumentsService {
     };
   }
 
+  createGatewayKubeDocument({ targetUri, origin }): DocumentGatewayKube {
```
Comment (gzdunek) [inline web/packages/teleterm/src/ui/services/workspacesService/documentsService/documentsService.ts:193]: Let's add types
```suggestion
  createGatewayKubeDocument({
    targetUri,
    origin,
  }: {
    targetUri: uri.KubeUri;
    origin: DocumentOrigin;
  }): DocumentGatewayKube {
```

File: web/packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts (hunk 11 above)
Comment (ravicious) [reply web/packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts:423]: Ok, I found it. https://github.com/gravitational/teleport/pull/27287#discussion_r1228154159

It looks like everything is according to what we discussed back then.

Yeah, it is updated in `refreshTitle`, but properly handling this requires also storing it on a document and a couple of other things, we can ignore this for now.

File: web/packages/teleterm/src/ui/services/workspacesService/documentsService/types.ts (hunk 10 above)
Comment (ravicious) [reply web/packages/teleterm/src/ui/services/workspacesService/documentsService/types.ts:140]: Ah, I forgot about that. 🤦 Let's just add the same comment above them to say they're tech debt, `DocumentGatewayCliClient` already has a comment like that.

File: web/packages/teleterm/src/services/tshd/types.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (ravicious) [reply web/packages/teleterm/src/services/tshd/types.ts:49]: Nice, I allowed myself to modify it a little bit so that we use `targetUri` as a fallback if all else fails.

During normal operation this shouldn't happen, but if there's a bug in tshd and we pass a wrong URI or something, it'll result in a better error message.

**Comment by ravicious** [comment]
> I have added some comments for removing the old implementation in [30bb201](https://github.com/gravitational/teleport/pull/29376/commits/30bb201d26f4b8e30240e9e57c86bea68a1a880f). I followed the way it's done in go but is there a standard how to mark deprecate in js? Also, most likely I have not covered everything that needs to be deleted later.

The comment you left should be enough. There's jsdoc's `@deprecated` which I added in a03a6625ad to a single function – if someone has their editor set up to integrate with the lang server, it should highlight the function at the callsites as deprecated. But those DELETE IN comments should be enough everywhere else.

File: web/packages/teleterm/src/ui/DocumentGatewayKube/DocumentGatewayKube.tsx [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+    await retryWithRelogin(ctx, doc.targetUri, () =>
+      ctx.clustersService.createGateway({
+        targetUri: doc.targetUri,
+        user: '',
+      })
+    );
+  });
+
+  useEffect(() => {
+    documentsService.update(doc.uri, { status: 'connecting' });
```
[the remark this answers is not in the document]
Comment (ravicious) [reply web/packages/teleterm/src/ui/DocumentGatewayKube/DocumentGatewayKube.tsx]: I pushed a commit which does that, as the semantics of using `useAsync` are somewhat complex and I didn't have a good example to link to.

**Review by ravicious — APPROVED** [review]

File: web/packages/teleterm/src/ui/DocumentGatewayKube/DocumentGatewayKube.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (greedy52) [reply web/packages/teleterm/src/ui/DocumentGatewayKube/DocumentGatewayKube.tsx]: thanks!

**Review by gzdunek — APPROVED** [review]
